### PR TITLE
DrivAerML: max-train-batches=788 (2x data/epoch)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

788 batches/epoch = 2x more car configurations per epoch than baseline (394). Each car configuration is seen twice per epoch instead of once. This tests whether more data diversity per epoch improves convergence, at the cost of fewer total epochs.

## Instructions

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 788 --max-eval-batches 200 \
  --wandb_group dm-throughput-sweep
```

## Reporting Requirements
- Best val `surface_rel_l2_pct` + epoch
- Total epochs run within time budget
- Test at best val checkpoint
- W&B run ID

## Baseline

**DrivAerML baselines:**
- val best: `3.997%` @ ep467 (PR #2898, W&B: `bht6h42t`, 4L/512d, AdamW lr=5e-4, T_max=30)
- test best: `6.244%` (PR #2648, 4L/320d, truncated eval — OLDER CONFIG)
- External targets: AB-UPT 3.82% | Transolver-3: 3.71%

Note: eren #2910 tested 788 batches at earlier config — results were unclear due to epoch budget. This tests at the current 4L/512d champion.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT kakashi:

## Results

### DrivAerML: max-train-batches=788 (2x data/epoch)

| Metric | This run | Baseline (PR #2898) | External targets |
|--------|----------|---------------------|------------------|
| Val best `surface_rel_l2_pct` | **11.57%** @ ep42 | **3.997%** @ ep467 | AB-UPT 3.82% \| Transolver-3 3.71% |
| Test `surface_rel_l2_pct` | **12.12%** | 6.244% (older config) | — |
| Total useful epochs | 128 (NaN at ep129) | 467 | — |
| Total epochs in budget | 321 (192 NaN) | 467 | — |

**Verdict: NEGATIVE.** 788 batches/epoch significantly underperforms the 394-batch baseline.

### Command
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 788 --max-eval-batches 200 \
  --wandb-group dm-throughput-sweep \
  --agent kakashi --wandb-name "kakashi/dm-train-batches-788"
```

### W&B
- **Run ID:** `ndxdsnkb`
- **Run URL:** https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/ndxdsnkb

### What happened

The core issue is that **788 batches/epoch interacts badly with the cosine LR schedule (T_max=30)**. With 2x batches per epoch, the cosine cycle completes in ~4 epochs instead of ~8, creating a fast warm-restart pattern.

**Three phases of training:**

1. **Epochs 1–42 (convergent):** Model made steady progress, reaching 11.57% val at ep42. Each cosine cycle trough delivered a new best. However, mid-cycle spikes appeared early (ep13: 54.8%, ep16: 47.4%) when LR peaked at 5e-4.

2. **Epochs 42–128 (oscillating/degrading):** The instability pattern worsened. Some cycles were smooth (ep20–29), but catastrophic spikes recurred (ep30: 52.5%, ep45: 46.3%). After ep45, gradient norms began escalating exponentially: 13 → 46 → 69 → 153 → 392 → ... The model partially recovered between spikes but never regained its ep42 best.

3. **Epochs 129–321 (NaN):** Gradient norms reached billions (ep128: 5.5 trillion) before hitting NaN at ep129. The remaining 192 epochs ran pointlessly at NaN until the 360-minute budget expired.

**Root cause:** With 788 batches/epoch, each epoch takes ~1.1 minutes vs ~0.55 minutes with 394 batches. The cosine T_max=30 operates on steps, so the LR cycle completes in half as many epochs. This creates abrupt LR transitions that destabilize training — especially the jump to peak LR (5e-4) every ~4 epochs. Without `--grad-clip`, these spikes are unconstrained and eventually cascade into permanent divergence.

**Why the baseline doesn't have this problem:** With 394 batches/epoch, the cosine cycle spans ~8 epochs, giving the model more gradual LR transitions. The baseline ran 467 stable epochs to reach 3.997%.

### Suggested follow-ups

1. **If 788 batches/epoch is worth retesting**, it needs `--cosine-t-max 60` (to match the ~8 epoch cycle length at 394 batches) and `--grad-clip 1.0` or `--grad-clip 2.0` to prevent gradient explosions.
2. **T_max should scale with batches/epoch** — the current config implicitly couples the LR schedule to the epoch length, which breaks when data throughput changes.
3. The NaN-but-keep-running behavior wasted ~3 hours of GPU time. A NaN early-stop check in the training loop would save significant compute.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3084 because ADVISOR: Closing — 788 batches/epoch is a dead end. Best val 11.565% at ep42 (+190% above 3.997% baseline), then catastrophic divergence at ep45 (grad norms reaching trillions). Root cause: T_max=30 operates on epoch count — with 2x batches/epoch the effective cosine cycle halves, creating more frequent LR peak shocks. Without gc there is no safety net. The 4x variant (#3051) showed the same failure. A viable follow-up is 788 batches + proportionally scaled T_max=60 + gc, but that's a fresh hypothesis. Good mechanistic analysis, kakashi..
